### PR TITLE
Add gre and qinq protocols

### PR DESCRIPTION
**Pull request by james-jra**
* Add support for basic GRE (generic routing encapsulation RFC1701).
* Add support for QinQ ethernet subtype (IEEE 802.1ad).
* Add some useful packet stack definitions

### CODE REVIEW

File: lua/proto/ip4.lua [hunk 1]
```diff
@@ -539,6 +542,7 @@ local mapNameProto = {
 	tcp = ip.PROTO_TCP, 
 	esp = ip.PROTO_ESP,
 	ah = ip.PROTO_AH,
+  gre = ip.PROTO_GRE,
```
Comment (emmericp) [inline lua/proto/ip4.lua]: please use tabs

**Comment by emmericp** [comment]
thanks! looks good.

just one questions: what about the ethertype? I thought that QinQ just adds another 0x8100 ethertype (which would be somewhat annoying to handle when parsing :/). I just read a little bit about QinQ and it seems like there isn't a standard followed by everyone :/

At least 0x88a8 (standard) and 0x9100 (juniper?) also seem to be used.

You are adding a cisco-specific ethertype as a constant, but your code later uses 0x8100 (hardcoded) default for both tags. Maybe add an option to make it configurable?

Can you please elaborate on your test setup: what devices did you test this with?


(I've luckily never had to use qinq and the last time I touched GRE is also a few years ago, we've migrated everything to vxlan :))

**Comment by james-jra** [comment]
I agree that QinQ should just add another 0x8100 ethertype, but there is no clear standard, just some "suggested" values.
It looks like only the ethertype of the outer 802.1Q header is variable.

I think the best option here is to remove the cisco-specific constant (I can't find where it came from in the first place anyway, I think it just turned up in a packet trace) and make the ethertype of the outer 802.1Q configurable. Is it worth extending this to the inner 802.1Q header too, even though it's not technically correct?

The downside here is you lose the ability to detect ethernet subtype automatically by checking ethertype against the list of constants.

As for the test setup, I'm afraid I can't go into specifics as it's an active project, but it does use some Cisco systems, hence the Cisco-specific ethertype

**Comment by emmericp** [comment]
thanks!

@scholzd will have a look at type-detection/parsing

**Pull request merged**